imagery/i.aster.toar/main.c

```diff
@@ -307,13 +307,13 @@ int main(int argc, char *argv[])
             }
         }
         for (i = 1; i <= MAXFILES; i++)
```

Comment:
I would change the for loop instead of its index. `for (i = 0; i < MAXFILES; i++)`.

Reply:
It seems that this is C code translated from Fortran code without taking care of the indices. E.g. MAXFILES is set to [15](https://github.com/OSGeo/grass/blob/master/imagery/i.aster.toar/main.c#L23), and outfd is an array of size MAXFILES. In https://github.com/OSGeo/grass/blob/master/imagery/i.aster.toar/main.c#L256 ff, the outfd array is set for indices 1 - 15 instead of 0 - 14, which can cause a segmentation fault because outfd[15] has never been allocated, only outfd[0] through to outfd[14].

The array indexing needs to be fixed first to avoid index out of array bounds errors.

Reply:
> E.g. MAXFILES is set to 15, and outfd is an array of size MAXFILES. In https://github.com/OSGeo/grass/blob/master/imagery/i.aster.toar/main.c#L256 ff, the outfd array is set for indices 1 - 15 instead of 0 - 14, which can cause a segmentation fault because outfd[15] has never been allocated, only outfd[0] through to outfd[14].

Yes this is what I meant to to address. (Please see commit message for suggested fix for each compiler warning type).

Big change re. to this seem to have committed with 
https://github.com/OSGeo/grass/commit/8661e3430d65af4acc76090ef92e5766a7ba6ad3.

Thanks for taking this up both of you, I should have brought this up to attention from the beginning, as I realised this might have real functional implications.

The question remain, where and how did the indices go wrong?

I’ll look into it, but I’m more than happy if someone beat me to it.  :)

Reply:
> The question remain, where and how did the indices go wrong?

I would like to understand that too. Is that likely that you translate it from one language to another and miss an obvious well-known difference? Sometimes people use some padding to make some loops easy and what not, but perhaps it was really just forcing Fortran indices on C.

Reply:
Set `int nfiles = 0;` in [L83](https://github.com/OSGeo/grass/pull/1248/files#diff-b73bcf9f5e3fdccbde25ae12cb5dcbb27d2af079a54944cb1994dd9a02d39d3fR83). 
The block from L229 to L243 should be
```
    for (; *ptr != NULL; ptr++) {
        if (nfiles == MAXFILES)
            G_fatal_error(_("Too many input maps. Only %d allowed."),
                          MAXFILES);
        name = *ptr;
        /* Allocate input buffer */
        in_data_type[nfiles] = Rast_map_type(name, "");
        infd[nfiles] = Rast_open_old(name, "");

        Rast_get_cellhd(name, "", &cellhd);
        inrast[nfiles] = Rast_allocate_buf(in_data_type[nfiles]);
        nfiles++;
    }
```
`nfiles--;` in L244 must then be removed.

The lines L279, L280 should be
```
        for (i = 0; i < MAXFILES; i++)
            Rast_get_row(infd[i], inrast[i], row, in_data_type[i]);
```
L309, L310 should be
```
        for (i = 0; i < MAXFILES; i++)
            Rast_put_row(outfd[i], outrast[i], out_data_type);
```
L312 - L317:
```
    for (i = 0; i < MAXFILES; i++) {
        G_free(inrast[i]);
        Rast_close(infd[i]);
        G_free(outrast[i]);
        Rast_close(outfd[i]);
    }
```
All array indices are now going from 0 to MAXFILES - 1. I could not add these suggestions directly to the code, thus here. The suggestions are untested.

Reply:
Updated a suggested.